### ISSUE TITLE
Potential bug: EXC_BAD_INSTRUCTION when using ImagePublisher

### ISSUE THREAD

**Issue by Nickersoft** [kean/Nuke#465]
Hey @kean,

Great work with Nuke! Been a long-time user. I recently made the switch over to using ImagePublisher + Combine in my codebase, and only recently started to experience a relatively weird issue. Upon opening my app, my images load, but then the app then crashes shortly thereafter on line 52:

https://github.com/kean/Nuke/blob/75da666fb8359f9cb954ed3168572c59c2383bd7/Sources/ImagePublisher.swift#L51-L53

Giving `EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)`. My guess is that the `subscriber` is null?

The XCode debugger shows nothing for the subscriber object:

![image](https://user-images.githubusercontent.com/2104129/117100914-ddb7b280-ad29-11eb-8502-88a7f2bcb6ee.png)

Yet the image is defined. I put some code in the `deinit` block of the class in which I'm using the publisher and the class is never deconstructed. Also, when using `print()` after loading the image, it shows the publisher as finishing prior to its crash. So unsure what's up. 

Curious if you or anyone else has experienced something like this. Can't post my code, as it's both complex and closed-source, but let me know if there's anything I can answer to help troubleshoot this!

**Comment by kean** [comment]
Hey, thanks for the report.

I haven't found any obvious issues after reviewing this component. I just added it in the latest version, and I can't say that I'm an expert in writing custom Publishers in Combine, so I could be missing something.

The memory cache lookup is performed synchronously as you subscribe. Please make sure you are not making assumptions about the publisher sending events asynchronously. Maybe some variable isn't initialized by the time the value is delivered?

I would appreciate it if you could post a simplified version of your code to help me investigate.

**Comment by Nickersoft** [comment]
Hey @kean, thanks for the quick response on this. Sorry it took me a bit to follow up. Mmm because I'm not sure of the cause, I'm not sure of how to simplify my code in a way that will effectively reproduce the issue. That said, after running my code a few times I realized I think it is being caused by a race condition. I did notice though that the crash seemed to go away after adding `.subscribe(on: RunLoop.main)` after subscribing to the image publisher... I think what might be happening is, because I jump around threads in my code, maybe the response was being returned on a different thread or something and that caused the issue?

So far adding an explicit subscription thread seems to work? Still not sure of why this would result in a BAD_INSTRUCTION, but it's my best guess.

**Comment by kean** [comment]
> I think what might be happening is, because I jump around threads in my code, maybe the response was being returned on a different thread or something and that caused the issue?

It sounds a lot like a usage issue. The method that creates a publisher is thread-safe, as far as I can tell. I'm going to close it if you don't mind. If you think there is an issue in `ImagePublisher` itself, please let me know.

**Comment by Nickersoft** [comment]
@kean I can confirm this issue is still occurring with the latest version.. and was actually able to create a minimal reproducible example :) Wasn't sure I'd be able to haha. I removed all my thread switching, so I don't think that's the issue – specifically, I'm noticing that the error occurs when you attempt to reload images from the cache. 

Steps to reproduce: 

1. Clone https://github.com/Nickersoft/NukeExample, install deps, run the app, etc.
2. Wait for the initial list of dog images to appear
3. Press "Reload"
4. The app should crash with an EXC_BAD_INSTRUCTION

From what I can tell, I think the address of the `subscriber` gets changed? Though nothing is ever deallocated, so I'm not sure what's going on. I tried structuring the example code as closely as I could to my actual production code.
